lib/xapi-stdext-unix/unixext.ml

```diff
@@ -156,21 +156,26 @@ let buffer_of_file file_path = with_file file_path [ Unix.O_RDONLY ] 0 buffer_of
 
 let string_of_file file_path = Buffer.contents (buffer_of_file file_path)
 
-(** Opens a temp file, applies the fd to the function, when the function completes, renames the file
-    as required. *)
+(** Write a file, ensures atomicity and durability. *)
 let atomic_write_to_file fname perms f =
-  let module Filenameext = Xapi_stdext_std.Filenameext in
-  let tmp = Filenameext.temp_file_in_dir fname in
-  Unix.chmod tmp perms;
-  finally
-    (fun () ->
-       let fd = Unix.openfile tmp [Unix.O_WRONLY; Unix.O_CREAT] perms (* ignored since the file exists *) in
-       let result = finally
-           (fun () -> f fd)
-           (fun () -> Unix.close fd) in
-       Unix.rename tmp fname; (* Nb this only happens if an exception wasn't raised in the application of f *)
-       result)
-    (fun () -> unlink_safe tmp)
+  let dir_path = Filename.dirname fname in
+  let tmp_path, tmp_chan = Filename.open_temp_file ~temp_dir:dir_path "" ".tmp" in
+  let tmp_fd = Unix.descr_of_out_channel tmp_chan in
```

Comment:
How about `tmp_chan`, seems like we're leaking it? I think we should close the `tmp_chan` instead of the file descriptor below, otherwise when the Gc finalizer runs on `tmp_chan` it will try to double-close the already closed file descriptor (at which point you either get an ENOENT error, or silently close the wrong file descriptor belonging to another file).

Reply:
That's... quite unexpected, there's no warning in `open_temp_file` not `descr_of_out_channel`. No mention of the behaviour of why the GC would close an `out_channel` but not a fd

Reply:
The fd is an integer `type file_descr = int`, so no finalizer can run on that.
Maybe you can submit a PR to the ocaml repository to describe the behaviour of `channel`? I think it is quite usual in OCaml that certain types close themselves when a finalizer is run.
And add a warning in the Unix module on the `file_descr` type, and the `close` function to say that OCaml doesn't provide any protection here, it is just as unsafe as the underlying `libc` call wrt to double close.